keras_nlp/models/falcon/falcon_backbone.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from keras_nlp.backend import keras
+from keras_nlp.layers.modeling.reversible_embedding import ReversibleEmbedding
+from keras_nlp.models.backbone import Backbone
+from keras_nlp.models.falcon.falcon_transformer_decoder import (
+    FalconTransformerDecoder,
+)
+
+
+class FalconBackbone(Backbone):
+    def __init__(
+        self,
+        vocabulary_size,
+        num_layers,
+        num_attention_heads,
+        hidden_dim,
+        # rope_scaling_factor,
+        layer_norm_epsilon=1e-5,
+        attention_dropout=0,
+        feedforward_dropout=0,
+        **kwargs,
+    ):
+        # Inputs
+        token_ids = keras.Input(shape=(None,), dtype="int32", name="token_ids")
+        # TODO: check falcon's padding.
+        # padding_mask = keras.Input(
+        #     shape=(None,), dtype="int32", name="padding_mask"
+        # )
+
+        # Embed Tokens
+        token_embedding_layer = ReversibleEmbedding(
+            input_dim=vocabulary_size,
+            output_dim=hidden_dim,
+            name="token_embedding",
+        )
+        x = token_embedding_layer(token_ids)
+
+        # Apply successive transformer decoder blocks.
+        for i in range(num_layers):
+            x = FalconTransformerDecoder(
+                num_attention_heads=num_attention_heads,
+                attention_dropout=attention_dropout,
+                feedforward_dropout=feedforward_dropout,
+                name=f"transformer_layer_{i}",
+            )(x)
+
+        x = keras.layers.LayerNormalization(
+            name="layer_norm",
+            epsilon=layer_norm_epsilon,
+        )(x)
+
+        sequence_output = keras.layers.Dense(
+            units=vocabulary_size, input_dim=hidden_dim, use_bias=False
+        )(x)
+
+        super().__init__(
+            inputs={
+                "token_ids": token_ids,
+            },
+            outputs=sequence_output,
+            **kwargs,
+        )
+        # All references to `self` below this line
+        self.vocabulary_size = vocabulary_size
+        self.num_layers = num_layers
+        self.num_attention_heads = num_attention_heads
+        self.hidden_dim = hidden_dim
```

Comment:
Before tests will pass, you will want a `get_config`, as well as `self.token_embedding = token_embedding_layer`. When we add unit tests later on that will catch this stuff.

Reply:
Added `get_config` and `token_embedding`.